Computer Craft integration for Update 6 Logistics Blocks

BirbIrl opened the pull request:
This is the backported version of @AsterAether, @Karotte128, @McArctic and I's efforts at bringing the logistics blocks to full compatibility with computercraft!
The original PR can be found here: https://github.com/Creators-of-Create/Create/pull/7669
The wiki PR is here: https://github.com/Creators-of-Create/wiki/pull/4
And there's also a minimal serial test world made specifically for this PR: https://github.com/BirbIrl/Create-CC-Integration-Tests
Note that 1.20.1 is technically the "base" branch for us, and our 1.21.1 branch merges these commits in with extra patches.

Oh also there's like 10 lines that got edited out cuz of linting (probably cause of me, i refuse to use an IDE and rock neovim just to feel something), whoops! I'm afraid of adding commits to undo that as to not confuse git with what should be merged and how between the branches. Hope that's okay!

oh and Aster, Karotte and Arcitc all have equal say here. if you have any questions and i'm unavailable when they are - take their word and assume i agree with them! :D

BirbIrl committed:
cherry-pick 1

BirbIrl committed:
walking around packager's actiavte() since it doesn't return a variable.
seems to work flawlessly.

BirbIrl committed:
i forgor to make StockTicker's .request() return an int

BirbIrl committed:
Made it so names in the enchant filter are optional

BirbIrl committed:
Made it so you can check enchantments by displayName (why not), made
enchantment checks exclusive, (one enchant per one set of enchant traits), made it so the number of enchants in the filter and items have to match. Will give a good example of it in the documentation linked in the pr

BirbIrl committed:
Packager now can get the num of items in attached storage

BirbIrl committed:
remove old comments because i'm silly

BirbIrl committed:
cherry-pick 2

BirbIrl committed:
cherry-pick 3

Karotte128 committed:
cherry-pick 4

BirbIrl committed:
cherry-pick 6

Karotte128 committed:
cherry-pick 5

BirbIrl committed:
made the getAddress and setAddress functions consistent accross
packager/postbox/frogge

BirbIrl committed:
"why isn't this variable what i told it to be?" - says programmer forgetting to set a variable t what they want it to be

AsterAether committed:
Add listPaymentInventory to StockTickerPeripheral

AsterAether committed:
cherry-pick 7

McArctic committed:
Added Configure method

Order is off when placing into requester

McArctic committed:
Updated Redstone Requester

fixed the ordering bug

AsterAether committed:
cherry-pick 8

BirbIrl committed:
rewrote the configure function, functionally the same (still broken)

McArctic committed:
Updated Redstone Receiver Peripheral

Added some safety stuff and made it actually save

BirbIrl committed:
Name = "minecraft:air"? {}? nil? no argument at all? Who needs to throw
an error? it's the same thing!

BirbIrl committed:
Lobotomized the requester

McArctic committed:
Update RedstoneRequesterPeripheral.java

Added a check to respect the redstone requesters default 256 max item limit

Karotte128 committed:
fix cherry-pick mistake

Karotte128 committed:
fix some things

BirbIrl committed:
As squid suggested, replaced IArguments with Optional<String> where
possible :D

Karotte128 committed:
repackager peripheral

Karotte128 committed:
add makePackage to Repackager

BirbIrl committed:
extra tablecloth functionality. half-working .setWares() included

Karotte128 committed:
somehow fix TableCloth item rendering on setWares

BirbIrl committed:
Removed a pointless comment and differencaited the .request functions
between stock ticker and redstone requester peripheral

BirbIrl committed:
made tablecloth setPriceTagItem be a void function like the rest

McArctic committed:
So, trying to just get this key but I cant seem to access it without jank

BirbIrl committed:
added the last feature: .getConfiguration() for the redstone requester

AsterAether committed:
Removed normal TableCloth as peripheral
Added ShopUpdate packet for updating shop rendering on computer updates

AsterAether committed:
Actually send the packet

AsterAether committed:
Actually actually send the packet

BirbIrl committed:
added back .isShop() because we love changing states

BirbIrl committed:
Epic air check that we forgor to pull

AsterAether committed:
Persisting addresses
- Persists the computer set address for Packager / Re-Packager
- Keep ephemeral nature of setAddress by clearing the hasCustomComputerAddress flag on attach / deattach
- Reorder Peripheral detection to account for Re-Packager being a subclass

SiffrinSuffrin commented:
Question: I may have missed it in the linked issue, but would this add the ability for computers to send item requests (similar to the existing stock-keeper) via a function (ie something like stockRequest(itemname, quantity, destination, source(optional, could just grab from any available storage)) and/or requestCraft(itemname, quantity, nameofporttosendmaterials)? Again, if this is already in #7669 or the PR, I apologise, thank you for your time.

edit - Nevermind, just checked the wiki linked in the issue, thanks for making this! This'll defo come in handy

lucyamonster commented:
Question: I don't even know if I'm allowed to ask this hear, So please excuse if I'm not. There isn't a way to get the address of a package at the moment right? If it's not will that ever be added? Thanks in advance.

BirbIrl commented:
> Question: I don't even know if I'm allowed to ask this hear, So please excuse if I'm not. There isn't a way to get the address of a package at the moment right? If it's not will that ever be added? Thanks in advance.

all good to ask here!
Sadly - not in a straightforward way.
Packages in packagers instantly get sucked in and turned into items. When the Event PR comes in from ElementW (https://github.com/Creators-of-Create/Create/pull/7453) we might be able to add an event to packagers that will return the package contents and it's address whenever it accepts a package :D

For now, if that doesn't get pulled in or our extra patch doesn't come through in time, you won't be able to tell what the package is addressed to unless you set a frogport to a specific address, then it'll call for those to be brought over and swatted out. That's one way to keep track of addresses in the current state of the PR.

The address of the package is hidden in NBT, so unfortunately there's no way to get the address with .getItemDetail() either, and we shouldn't inject into an existing Computercraft function.

There was an idea to make the repackager be able to analyse packages in the inventory it's connected to, and assemble them via code, but it turned out to be a project wayyy over our heads, considering the amount of things we've already added. We're a bit out of steam.
Anyone is free to try and contribute to our fork (which will get synced with this PR), though! We have a group on discord regarding this PR if anyone's interested in contributing. Anyone is free to join, just add me @birbirl on discord :>

TropheusJ commented:
the original PR should be closed, this one will be merged up if accepted

BirbIrl commented:
> the original PR should be closed, this one will be merged up if accepted

On it!
Do note though that the other PR has the additional patches already included for your convenience to merge up to 1.21.1 cleanly. the way peripherals are registered has changed between versions, as well as some rendering stuff. So feel free to look at the 1.21.1 branch of our fork when merging up. Thanks!

lucyamonster committed:
Adding `getPackageAddress()` to re-packagers.

lucyamonster committed:
'getPackageAddress()' now returns 'nil' when there's no package.

lucyamonster committed:
fixing a couple bad imports in 'RepackagerPeripheral.java'

BirbIrl committed:
Merge pull request #1 from BirbIrl/mc1.20.1/review-fixes

Review Fixes

BirbIrl committed:
Merge pull request #2 from lucyamonster/mc1.20.1/dev

`getPackageAddress` function for re-packagers

BirbIrl committed:
Changed "checkPackage" to "getPackageItems" and gave it to the
repackager as well

BirbIrl commented:
Small set of tweaks came in:
1: the Repackager can now get the package's address with `.getPackageAddress` (Thank you, lucyamonster!)
2: the checkPacakge() function has been renamed to `.getPackageItems`, the repackager now also has access to that function. 
3: the packager and repackager now remember their address when unloaded, but still forget it if they call a package with a computer detached. (Thank you, AsterAether!)

Gonna update the wiki pr with the new information.

also we're no longer maintaining the 1.21.1 branch on our end. :D

BirbIrl committed:
me when i forget about the includes

ThePuzzlemaker commented:
Would it be possible to add a filtered `makePackage()` function to the packager and/or to have multiple filters for the stock ticker's `requestFiltered()` function?

BirbIrl commented:
> Would it be possible to add a filtered `makePackage()` function to the packager and/or to have multiple filters for the stock ticker's `requestFiltered()` function?

for packager's filtered `makePackage()` - that is not a base create feature. the packager grabs the first few items from an inventory and packages them.
The stock ticker supports sorting items by tags in the ui, hence why it excluisvely gets the `requestFiltered()` function.
The point of the CC additions is to automate manual tasks (like changing signs for addresses, powering it with redstone, changing the redstone requester's filter, changing shop prices), not add brand new functionality to blocks, which all serve specific purposes.
The previous CC integration added the ability to automate changing the speed of the rotational speed controller - but it doesn't let you set decimal speeds, it doesn't let you set 0rpm either. all of which it *could* do, but the first one is against Create as a whole, and the other is the Clutch's task.
Hope that makes sense.
Technically we upgraded the shop's functionality because you can change wares without placing/breaking the cloth, but i think it's fine.
The only block we're ***looking*** at "expanding" functionality is the re-packager, but we feel like the PR is in a healthy spot without that.

Finally, as for multiple filters in the `requestFiltered()` function, just call the function multiple times with different filters. If we were to add multiple filters on the java end, it'd just be us looping through every arg - so instead of that, we let people run the function multiple times in lua :D.

lucyamonster committed:
`getPackageAddress()` now uses a different way of reading addresses.
And adding the `setPackageAddress()` function to packagers and re-packagers.

lucyamonster committed:
`setPackageAddress` returns `true` when:
Provided with a string after renaming a package
Provided no string after setting package name to `""`.  Not visible when hovering over the item with cursor.
returns `false` when:
no package found.

BirbIrl committed:
Merge pull request #3 from lucyamonster/mc1.20.1/dev

`getPackageAddress()` now uses a different way of reading addresses.

aearil committed:
Merge remote-tracking branch 'origin/mc1.20.1/dev' into cctweaked-logistics

aearil committed:
Fix redstone requester

Redplcs commented:
Hello guys. When this PR will be finished, what would happen with it? Will it come into newest major version or minor? Or it will be as separate extension mod? I don't really know how your's development pipeline happens

McArctic commented:
> Hello guys. When this PR will be finished, what would happen with it? Will it come into newest major version or minor? Or it will be as separate extension mod? I don't really know how your's development pipeline happens

Depends, its up to the create devs if they want to merge it into the mod or not. I cannot speak to what version it will come into either if it gets added at all

IThundxr commented:
> Hello guys. When this PR will be finished, what would happen with it? Will it come into newest major version or minor? Or it will be as separate extension mod? I don't really know how your's development pipeline happens

Currently feature work is on the back burner while bugs are fixed, once feature work is resumed any and all feature PRs will be reviewed.

Redplcs commented:
> Depends, its up to the create devs if they want to merge it into the mod or not

So what if devs will discard PR. Will it become as separate mod?

McArctic commented:
> > Depends, its up to the create devs if they want to merge it into the mod or not
> 
> So what if devs will discard PR. Will it become as separate mod?

No clue lol, I only did a lil bit for this pr and setup some ground work stuff with like one block lol. So I guess it depends what the rest of the group wants to do. Altho I have a good feeling it will get merged eventually:)

Karotte128 commented:
> > Depends, its up to the create devs if they want to merge it into the mod or not
> 
> So what if devs will discard PR. Will it become as separate mod?

If this PR is not merged we can look into making this a addon. But I hope this will be merged. I do not want to waste time creating the addon if this ends up getting merged.

Redplcs commented:
> Altho I have a good feeling it will get merged eventually:)

I don't think this is a good idea to patch few lines of code just for better integrations with other mods. Having patches for individual mods in one day will become a mess that you must to support. That will increase the complexity and maintenance will be must harder. If the integration more relies on reflection then it's much worth, because in term of modding when one mod getting methods of other mods through reflection someday it will break the backward compatibility, because any refactoring will break everything.

I think addon will be more realistic here. The developers doesn't need to support the patch. I assume they think the same

IThundxr commented:
> > Altho I have a good feeling it will get merged eventually:)
> 
> I don't think this is a good idea to patch few lines of code just for better integrations with other mods. Having patches for individual mods in one day will become a mess that you must to support. That will increase the complexity and maintenance will be must harder. If the integration more relies on reflection then it's much worth, because in term of modding when one mod getting methods of other mods through reflection someday it will break the backward compatibility, because any refactoring will break everything.
> 
> I think addon will be more realistic here. The developers doesn't need to support the patch. I assume they think the same

All the code used is part of CC: Tweaked's API, and create already has CC: Tweaked compat, a addon is not the realistic approach here.

Redplcs commented:
> All the code used is part of CC: Tweaked's API, and create already has CC: Tweaked compat, a addon is not the realistic approach here.

Wait what? Why the hell Create depends on CC: Tweaked? What's the purpose of it??

Vap0r1ze commented:
> > All the code used is part of CC: Tweaked's API, and create already has CC: Tweaked compat, a addon is not the realistic approach here.
> 
> Wait what? Why the hell Create depends on CC: Tweaked? What's the purpose of it??

It doesn't it just uses it's API when it's present. Aka it's "compatible"

Redplcs commented:
> It doesn't it just uses it's API when it's present. Aka it's "compatible"

Through reflection? I still don't see any reasons for relying on other mod, that's not a part of the core, even if it's presented on fly. Anyway you know way better than me, so good luck with your project

BirbIrl commented:
> So what if devs will discard PR. Will it become as separate mod?

All of our work is under MIT! i don't think any of us mind if our code gets used in any addon willing to add the compatibility as long as we get credit!
`ComputerCraft Create Bridge`/`Create Tweaks and Additions` both already add some things we don't have in the base mod regarding computers, so the pr could be adapted to both those mods and sent their way.
I personally won't be making any other pr's or making my own addon - but anyone is free to! I'd advise waiting for the verdict from the create team - and just compiling our fork if you wanna mess around with the features.

> I still don't see any reasons for relying on other mod

tldr: [It allows people to make cool things](https://www.reddit.com/r/CreateMod/comments/1atxbko/discrete_tree_farm_w_computercraft/), while still staying with create's base design principles. Also not using the CC api would really just be reinventing the wheel, and after dealing with minor lua vs java types and type safety i can say that squid did an amazing job on the api and i wish upon no one else the pain of making that infrastructure ever again.

Funnily enough using the computercraft api has actually led to issues already with a crash on boot, but it only happened if you have both `computercraft: tweaked` and `create` installed. That got patched in like 2 days on CC: Tweaked's end (thanks Squid!). It's worth it, even if it leads to some extra work sometimes.

And yeah we're a patient bunch, take all the time you guys need to squash bugs :D

aearil commented:
Hi, a bit of test feedback for you here: we've been running your PR (updated for 6.0.4, see https://github.com/aearil/Create/tree/cctweaked-logistics) for nearly a month now without any issues.

In particular my GF built a 1500 LoC logistic system heavily relying on the redstone requester's `setAddress()`, `configure()`, and `request()` APIs, as well as the stock ticker's `list()`. That feels like a pretty good stress test (that computer basically replaces and improves upon a full wall of gauges).

The only caveat here is that I didn't take the time to get crafting info embedded into packages working with the new `PackageOrderWithCrafts` API.

In any case, let me now if you want me to bring my changes to this PR.

BirbIrl commented:
> In any case, let me now if you want me to bring my changes to this PR.

Anyone is free to contribute! Development on this is "done", as in everything we felt like adding - we added, but we have a group chat on discord you can join by adding me (@birbirl on discord) where we can talk about if adding something is a good idea or not before spending a day implementing something only to realize it wasn't a good idea to begin with (also we can add you to our fork as a contributor).

Also feel free to PR your updated version on our fork so that the create devs have an easier time, since you seem to have also added a patch for the redstone requester (was it not working on 6.0.4 without it?)

and yeah the [PackageOrderWithCrafts](https://github.com/Creators-of-Create/Create/blob/6560a3e52f812b6742b77486e4d5ce0e6a933a2d/src/main/java/com/simibubi/create/content/logistics/stockTicker/PackageOrderWithCrafts.java#L20) api makes as much sense in my tiny bird brain as lovecraftian horrors. If you feel like working with it, go ahead!

Also since your partner had a lot of experience with the pr, mind asking her how was the [documentation](https://github.com/Creators-of-Create/wiki/pull/4)? anything missing, unclear, maybe in need of examples?

Thanks :D

aearil committed:
[CC:Tweaked] Fix RedstoneRequesterPeripheral crafting metadata

Commit:
RedstoneRequesterPeripheral: add configureCraft for mech crafter compat

BirbIrl committed:
Merge pull request #4 from aearil/cctweaked-logistics

Bump CC:Tweaked logistic support to 6.0.4 and split redstone requester's `configure` function to match new version's standards

BirbIrl commented:
Hey @IThundxr, hope you don't mind the ping after the couple of months of this PR - but we reached a game-design choice we can't quite decide on.

So, originally this PR allowed people to make crafting requests with the redstone requester by adding blank spaces, but some time between 6.0.1 and 6.0.4 this has become impossible via a normal request, so now as a "bandaid" fix we let a computer set a requester's configuration to a crafting request with `configureCraft`, and normal requests with `configure`. this means, if you interact with the ui - it gets converted to a normal request. I.e certified jank teritorry. (As well as, this now gives the redstone requester a function it previously never had, which we try to avoid in this pr.)
We believe this is... *barely passable*. You can check the documentation on how it works [here](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/redstone-requester#configure).

So, right now the redstone requester can `.setAddress` then `configure` or `configureCraft` and finally `.request` what was assigned in the `configure`/`configureCraft`'s 9 slots 
While the stock ticker just has access to `requestFiltered`, which takes an address and filter, then goes through the filter and requests by enchants, durability and other traits CC can read. Stock ticker also has the `list` and `listDetailed` functions which add a way to look into the items in the storage as a lua table. 

But we wanna make it as good as it can be, like the rest of this pr, so we have a couple options:
- We can turn the `setAddress`+`configureCraft`+`request` combo from the redstone requester and put it on the stock ticker, as something like `requestCraft`. This means the stock ticker would `requestCraft` and `requestFiltered`, while the redstone requester is `configure`-d for normal package requests
- We can do the above, and also give it the `request` to work like the redstone requester's `setAddress`+`configure`+`request` combo, "because it makes sense", but then there's no reason to use the redstone requester, because the stock ticker will have all of it's functionality basekit (except i guess the ability to be mutable, which is cool for the redstone requester). This has been implemented by Glass on Aearil's fork, [here](https://github.com/aearil/Create/commit/b5f7ab87244d3fd9aa3291390b93f618bf9b750b)
- We can have it how it is now (redstone requester gets to `configure` and `configureCraft`), and possibly lock out the stock ticker's ui while it's connected to a computer, like the sequenced gearshift, so players can't break the request recipie. This is still a design break where the requester can make crafting recipes, which recent patches removed
- We could possibly make a second pr that normalizes and adds a switch to the redstone requester to toggle between making crafting and normal requests, but then that slowly gnaws at stock gauge's functionalities

Sorry for blasting like days worth of discussion between us upon ye, but getting a thumbs up on any of the ideas (or leaving as is) would help us a ton :3

ps. we're working with powerpig now on combining [their pr](https://github.com/Creators-of-Create/Create/pull/8390) with ours :D

IThundxr commented:
> Hey @IThundxr, hope you don't mind the ping after the couple of months of this PR - but we reached a game-design choice we can't quite decide on.
> 
> So, originally this PR allowed people to make crafting requests with the redstone requester by adding blank spaces, but some time between 6.0.1 and 6.0.4 this has become impossible via a normal request, so now as a "bandaid" fix we let a computer set a requester's configuration to a crafting request with `configureCraft`, and normal requests with `configure`. this means, if you interact with the ui - it gets converted to a normal request. I.e certified jank teritorry. (As well as, this now gives the redstone requester a function it previously never had, which we try to avoid in this pr.) We believe this is... _barely passable_. You can check the documentation on how it works [here](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/redstone-requester#configure).
> 
> So, right now the redstone requester can `.setAddress` then `configure` or `configureCraft` and finally `.request` what was assigned in the `configure`/`configureCraft`'s 9 slots While the stock ticker just has access to `requestFiltered`, which takes an address and filter, then goes through the filter and requests by enchants, durability and other traits CC can read. Stock ticker also has the `list` and `listDetailed` functions which add a way to look into the items in the storage as a lua table.
> 
> But we wanna make it as good as it can be, like the rest of this pr, so we have a couple options:
> 
> * We can turn the `setAddress`+`configureCraft`+`request` combo from the redstone requester and put it on the stock ticker, as something like `requestCraft`. This means the stock ticker would `requestCraft` and `requestFiltered`, while the redstone requester is `configure`-d for normal package requests
> * We can do the above, and also give it the `request` to work like the redstone requester's `setAddress`+`configure`+`request` combo, "because it makes sense", but then there's no reason to use the redstone requester, because the stock ticker will have all of it's functionality basekit (except i guess the ability to be mutable, which is cool for the redstone requester). This has been implemented by Glass on Aearil's fork, [here](https://github.com/aearil/Create/commit/b5f7ab87244d3fd9aa3291390b93f618bf9b750b)
> * We can have it how it is now (redstone requester gets to `configure` and `configureCraft`), and possibly lock out the stock ticker's ui while it's connected to a computer, like the sequenced gearshift, so players can't break the request recipie. This is still a design break where the requester can make crafting recipes, which recent patches removed
> * We could possibly make a second pr that normalizes and adds a switch to the redstone requester to toggle between making crafting and normal requests, but then that slowly gnaws at stock gauge's functionalities
> 
> Sorry for blasting like days worth of discussion between us upon ye, but getting a thumbs up on any of the ideas (or leaving as is) would help us a ton :3
> 
> ps. we're working with powerpig now on combining [their pr](https://github.com/Creators-of-Create/Create/pull/8390) with ours :D

I think leaving it as is for now is fine, if there's any changes that need to be made later we can go ahead and do those based on what the rest of the team thinks about the different methods

MaxOrtGit committed:
Packager Peripheral Update

Added many new functions to the packager and a bit updated/removed

MaxOrtGit committed:
Package Detail Providers

MaxOrtGit committed:
Updated API

MaxOrtGit committed:
Remove vscode

MaxOrtGit committed:
Allow for creating PackageLuaObject without block

MaxOrtGit committed:
Updated RePackager

MaxOrtGit committed:
Remove getItemCount for API consistency

MaxOrtGit committed:
Added more functions for order data

rmortes commented:
Hey! Mind if I add a bit of feedback for this issue as a player, and not a developer?

I would love this integration for the purpose of creating a pseudo ME network using just Create and CC. The main features I'd like to replicate are autocrafting, crafting queue and wireless access.

Wireless access is probably impossible with just Create and CC. I'd need to add an Ender Chest+ mod. Although, I could load a turtle full of packages and send it to my location... But I digress, as I was saying, this feature has nothing to do with this PR.

Autocrafting is another feature that I'd like to implement using CC. I know that stock gauges already implement autocrafting, but I was thinking about something more on-demand. Instead of needing to have some stock to craft things, I'd like to request a dispenser and have my system craft some sticks, wait for the sticks, craft a bow, wait for the bow, and then craft a dispenser. For this I think I'd need a way to interface with the network's promises. I don't know if this PR's .requestCraft method creates a Promise in the network, since I don't quite know how the Promises work code-wise, but if so I'd love to be able to hold a reference to the promise in my CC computer and to be able to query for the promise status.

Now, what I mean for "crafting queue" is this menu AE2 has in the terminals that let you know the current state of your crafting orders. I'd love to have a big screen with the current status of my factory: This autocraft has been ongoing for X seconds and it's waiting for sticks, this stock ticker has requested 64 cobblestone to be smelted, etc. For this I think I'd need to be able to access Promises. Maybe to be able to query for long a Promise has existed, and data regarding the stock gauge that created the promise.

As a bonus, if crafting orders create Promises, and that promise had the crafting recipe, maybe the system could automatically learn crafting patters like Isy's Inventory Manager does for Space Engineers.

I don't know if what I'm asking for is technically feasible, so if I'm talking crazy here forget me!

BirbIrl commented:
> Hey! Mind if I add a bit of feedback for this issue as a player, and not a developer?
> I would love this integration for the purpose of creating a pseudo ME network using just Create and CC. The main features I'd like to replicate are autocrafting, crafting queue and wireless access.

tldr: you can do it right now in lua, with the functions provided, so no need

the long version:
i hear you! All of this is actually entirely possible within lua code right now! As per the rules of Computercraft - you've got to code it yourself (or copy someone else's code :P.). The point of peripheral apis is to allow computers to access basic functions of a block, to then, using the api, make your own complex systems. Using the redstone requester you can already make packages for the autocrafter, and you can route it however you want with the existing create address system (also modifiable with computers).

Would implementing it as basekit with the PR make it more convenient for everyone? yeah, but there's a reason why computercraft turtles don't have a "craft" function, that just lets you pick 9 slots in whatever order and crafts them - instead we have dedicated 9 slots inside the inventory that, unless carefully worked with, will clog up after a couple crafts, as whatever you craft is dumped into the same inventory.
There's a reason why there is no "getPos" function. You first have to place 3 turtles with modems and make your own trigonometric gps system. CC:Tweaked only provides a [helper program](https://tweaked.cc/module/gps.html) for it, but it's entirely written in lua, you can code your own!

I think we should take notes from the way mekanism implements the computercraft integration. it lets you manage your upgrades, machine settings, but doesn't suddenly let you make the machines do more than they were designed. They simulate things the player would do with the machines, and let you automate that.

At most, i'd allow computers to manage a wall of stock gauges - but coding that is beyond me and everyone else involved. Plus you'd still have to have turtles placing the stock gauges, et cetra. no idea how that'd work. honestly. The stock gauges source code scares me.

So, we're providing all the building blocks needed to make whatever you want. It will take time and effort - and that's the fun part! Why make an entire hard-coded-java autocrafting-queue stack, that half the people don't like, and would want to work slightly differently _("ermm i encode my addresses like "destination.reicpieid", not "destination!recipeid" because it conflicts with the rest of my address system, please add a string parameter in .makeQueuedRequestOrSomething() to change the separator")_ . Plus some people would want to sort packages via forgports and chians, some would wanna sort them via turtles - et cetra.
I think giving the player versatile tools to make specialized systems is *the* create way of doing things. Anything else would be room for an addon.

Oh and i'm speaking for myself here, not as the whole team working on the PR

MaxOrtGit committed:
CC Generic API Alignment

 Changed Postbox, Packager (conflicts with other PR), Re-Packager and Stock Ticker to no longer have getItemCount and listDetailed and instead have getItemDetail for performance

Removed optional for setAddress for functions that nil is the same as an empty string. This is so nil doesn't mean different things for Packagers and the rest

Added list and getItemDetail to Frogport and Postbox

MaxOrtGit committed:
Everything but count

MaxOrtGit committed:
Added max count to stock ticker

MaxOrtGit committed:
Added a way to compare lua objects to tables

MaxOrtGit committed:
Package object given in getDetails

MaxOrtGit committed:
LuaComparable for Packages

MaxOrtGit committed:
Merge pull request #7 from MaxOrtGit/mc1.20.1/CCStandardStockTicker

Made stock ticker requesting work with any arbitrary data from item details

Electric131 commented:
> Finally, as for multiple filters in the `requestFiltered()` function, just call the function multiple times with different filters. If we were to add multiple filters on the java end, it'd just be us looping through every arg - so instead of that, we let people run the function multiple times in lua :D.

Just stumbled on this PR recently, so I'm sorry if this reply is a little old xD. I already cloned this and have been messing around a bit but I can't wait to properly see what's possible with this, but I'd like to mention I disagree with this idea of calling requestFiltered multiple times.

The first thing I noticed was the documentation was a bit out of date, which is fine since it was relatively easy to figure out the new format. However, the issue I have is with calling the `requestFiltered()` function multiple times, since it results in each item being sent as a separate package. I can see how it would be useful if it stacked in some kind of cache, similar to how the DisplayLink works with `write()` and `update()`, but I really want to see a form of multiple request system, in any implementation.

I went ahead and took a stab at it, even if it's just for my personal use, and rewrote it into the original `request()` function that takes an array of items to process them into one request. From my naive understanding, it should be about the same speed as calling `requestFiltered()` multiple times, except it has the advantage of packing all the items into one package and will not send the package if any of the requests fail.

If anyone is interested in seeing my implementation, I have the fork on my github. Although it's forked from BirbIrl's fork so I don't think it can merge but it doesn't seem like this is the approach Birb is going for anyway.
(The forking thing is just because I'm not experienced enough with git to PR a PR..)

Edit: I did find out how to merge my fork into Birb's fork in the Github UI, it's just something I've never done before I guess. So if anyone is curious (or if Birb/others want to merge it with this) it's there if you want it

BirbIrl commented:
mr. create mainter a 9th contributor has hit the 7k lines changed "minor mod integration" pr

Noted, it's 5am for me and i'm about to head to bed, but i see this, @Electric131, and sent a screenshot to the group chat and @MaxOrtGit (who's most recently been working on the filter logic, and just generally been doing god's work) is already looking through it. At first glance, i agree, getting 1 request for multiple filters sounds like a good idea, but i'm looking at the code and my brain is blanking. I feel like there's probably a cleaner way to do this. I trust @MaxOrtGit in figuring out the approach (Thank you, both!).

Personally i've been completely out of gas on this, plus i'm a little terrified that we're at +7k -2k lines of code in this pr (going from like, +1000, -2) so there must've been sooome linting/rebase incident, but if anyone wants to do the janitor work/review the existing pr's it'd be much appreciated :3

Either way i'm going back to crawling at a snail's pace with this pr as my finals approach for the next month or so :(

Oh and to the create maintainers: If you're done fixing all the bugs and are open to feature pr's again, let me know. I could just dedicate a day to locking in and implementing, cleaning everything and updating the wiki so that reviewing and merging this is as nice as possible. It'd probably be enough motivation for my brain. Knowing it'd be the last cleaning up i'll have to do before passing alllll the responsibility down to main branch nyeheheh >:)

MaxOrtGit committed:
Added custom operators with _op

You can specify the operator by using a map with the keys _op and value
so for getting items were there are more than 64 in the network you would  specify
`{ count = { _op = ">", value = 64} }`

Operators are:
- not: true if value does not pass filter
- any: true if any items in value pass filter
- all: true if all items in value pass filter
- type: sees if type is nil, number, string, boolean, table, list, map, or object (LuaComparable)
- number comparison: >, >=, <, <=, ==, and ~=
String matching:
- glob: same as package matching
- regex: regex comparison

```lua
sk.requestFiltered({
  -- Get all names that are either:
  -- minecraft
  -- don't contain a w
  -- a crushing wheel
  name = { _op = "any", value =
    {
      {_op = "glob", value = "minecraft:*"},
      {_op = "not", value =
        {_op = "glob", value = "*w*"}
      },
      "create:crushing_wheel"
    }
  },
  -- Get items in stockTicker that have a count between 2 and 100
  count = { _op = "all", value =
    {
      { _op = ">=", value = 2 },
      { _op = "<=", value = 100 }
    }
  },
}, "chest_2")

```

MaxOrtGit committed:
Small tweaks

MaxOrtGit committed:
Added LuaBigItemStack so count will be accurate

MaxOrtGit committed:
Merge pull request #9 from MaxOrtGit/mc1.20.1/CCStockTickerOtherOperations

Request filtered special comparison with _op

simibubi commented:
Hi, thanks a lot for the contributions.

I tried merging this to a local branch and the diff looks ok (+1311-8), must be github acting up

Happy to merge this as soon as you consider it ready.

I will warn you about messing with PackageOrderWithCrafts information, though. The re-packager does not validate it on its end and will dupe or void items if the recipe and order data does not match up.
This has been a common source of issues even for base create.

BirbIrl committed:
syncing with upstream
Merge remote-tracking branch 'remotes/upstream/mc1.20.1/dev' into mc1.20.1/dev

BirbIrl committed:
Merge pull request #5 from MaxOrtGit/mc1.20.1/packagerExtension

Packager Peripheral Update

BirbIrl committed:
Merge branch 'mc1.20.1/dev' into mc1.20.1/CCGenericInventory

BirbIrl committed:
Merge pull request #6 from MaxOrtGit/mc1.20.1/CCGenericInventory

CC Generic API Alignment

BirbIrl committed:
pacthes post merge of #5 and #6

BirbIrl committed:
cleanup after packager rework

BirbIrl committed:
Fixed a bug that made it so .getAddress() would just lie to you, and
would only give an accurate address after you order a package. Had to
make `updateSignAddress` in PackagerBlockEntity public tho :(

BirbIrl committed:
first half of redoing .list and .getItemDetails to be more like CC:T

BirbIrl committed:
Second part, reverting .list() and .listDetailed() for the stock ticker
peripheral for sanity reasons

Karotte128 reviewed (changes requested):
The .factorypath file looks like it should not be pushed to git. @BirbIrl Did you forget to add it to gitignore?

BirbIrl commented:
> The .factorypath file looks like it should not be pushed to git. @BirbIrl Did you forget to add it to gitignore?

your honor i plead an oopsie daisy

BirbIrl committed:
whoopsie daisy nixos moment, also simplified PackageLuaObject's code and
added to note to the PackageOrderLuaObject code

BirbIrl committed:
updated .gitignore to include nixos weirdness

BirbIrl committed:
nvm it's not a nixos thing, it's just a universal gradle moment

BirbIrl committed:
removed an include from PackageLuaObject and linted

BirbIrl committed:
added packager getAddress tweak to repackager

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/RedstoneRequesterPeripheral.java:
```diff
@@ -0,0 +1,131 @@
+
+public class RedstoneRequesterPeripheral extends SyncedPeripheral<RedstoneRequesterBlockEntity> {
+
+	// private final ScrollValueBehaviour targetSpeed;
```
targetSpeed unused

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/RedstoneRequesterPeripheral.java:
```diff
@@ -0,0 +1,131 @@
+
+	public RedstoneRequesterPeripheral(RedstoneRequesterBlockEntity blockEntity) {
+		super(blockEntity);
+		// this.targetSpeed = targetSpeed;
```
again, targetSpeed unused

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/RedstoneRequesterPeripheral.java:
```diff
@@ -0,0 +1,131 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
```
import seems to be useless

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/FrogportPeripheral.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import dan200.computercraft.api.detail.VanillaDetailRegistries;
```
unused import

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/RepackagerPeripheral.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.Optional;
+
+import com.simibubi.create.content.logistics.packager.repackager.RepackagerBlockEntity;
+import com.simibubi.create.content.logistics.stockTicker.PackageOrderWithCrafts;
```
unused import

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/RepackagerPeripheral.java:
```diff
@@ -0,0 +1,80 @@
+		if (!blockEntity.heldBox.isEmpty())
+			return false;
+		blockEntity.activate();
+		if (blockEntity.heldBox.isEmpty())
```
why this double if instead of else? Also, when is `return true;` ever reached??

Karotte128 replied:
Finally, I understand what I did there

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/StockTickerPeripheral.java:
```diff
@@ -0,0 +1,153 @@
+import java.util.ArrayList;
+
+import dan200.computercraft.api.lua.LuaFunction;
+import net.minecraft.world.item.ItemStack;
```
unused import

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/StockTickerPeripheral.java:
```diff
@@ -0,0 +1,153 @@
+
+public class StockTickerPeripheral extends SyncedPeripheral<StockTickerBlockEntity> {
+
+	// private final ScrollValueBehaviour targetSpeed;
```
unused variable

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/StockTickerPeripheral.java:
```diff
@@ -0,0 +1,153 @@
+
+	public StockTickerPeripheral(StockTickerBlockEntity blockEntity) {
+		super(blockEntity);
+		// this.targetSpeed = targetSpeed;
```
unused variable

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/StockTickerPeripheral.java:
```diff
@@ -0,0 +1,153 @@
+	}
+
+	/*
+	 * for every item in the netowrk, this will compare that item to the CC args
```
typo in network

Karotte128 commented on src/main/java/com/simibubi/create/compat/computercraft/implementation/peripherals/StockTickerPeripheral.java:
```diff
@@ -0,0 +1,153 @@
+		PackageOrder order = new PackageOrder(validItems);
+		blockEntity.broadcastPackageRequest(RequestType.RESTOCK, order, null, address);
+
+		/*
```
unused code

Karotte128 reviewed (changes requested):
this is all I found.
Formatting is a bit off at some places, I might fix it if you want @BirbIrl.

Karotte128 committed:
formatting + remove unused imports + small fixes

BirbIrl committed:
epic grammar fix

BirbIrl commented:
Alright so i want to make sure this pr is as best as it can be. I've been making a serial test world for all computercraft x create stuff, which, while developing it, helped me find two bugs already (that sprung up while updating the branch).

I feel like for the quality of the create mod it's not a good idea to pull in the pr just yet until i get done with that, which, was supposed to take a week, but could also take weeks.

After i'm done making the world i'll try and either make it work as a schematic somehow(?) or upload it and the scripts so that every time there's a create update, checking if there's any changes on the CC end will be easy.

Also the stock ticker's requestFiltered() function i'm not entirely happy with, and i wanted to implement @Electric131's suggestion of grouping orders

Other than that i have a couple of vanity things i wanted to change but stuff came up for me this weekend sooo until then i'm making this pr a draft for the time being. I don't want people to go crazy adding comptuercraft scripts for the logistics chains only to find out next update there's a breaking change :>

> I will warn you about messing with PackageOrderWithCrafts information, though. The re-packager does not validate it on its end and will dupe or void items if the recipe and order data does not match up. This has been a common source of issues even for base create.

Should be fine! we only really fetch information of the order, never change it after the first time the package is made and we're not planning on implementing that either :3.

MaxOrtGit committed:
configureCraft allows for just string

BirbIrl committed:
.listDetailed() removed, instead we now use .list(?Detailed)

BirbIrl committed:
refactored .requestFiltered and added _requestCount functionality

BirbIrl committed:
refactored stock ticker's list, getItemDetail and paymentInventory ()'s

BirbIrl committed:
summer cleaning

BirbIrl committed:
renamed stock ticker functions as per vote

BirbIrl commented:
Passed my finals, did some tweaks, here's a status update:
- The PR is now officially feature complete. Any new commit should be a bugfix/cleanup.
- The wiki PR is ***very*** outdated and i'll be working on it this week
- I'm still working on my serial tests, which already squashed one bug, so there's a good chance i'll find another by the time i'm done

Future plans we'd love to implement (well), but didn't get to:
- implement the generic inventory api handler for the stock ticker, so we can export items from it. Currently held back by jank on CC's end on adding `generic` capabilities to `IPeripheral`'s and nesting functions and general not-worth-the-effort-ness on my end, since it's possible, would just be a lot of rewriting. Related topic here https://github.com/cc-tweaked/CC-Tweaked/discussions/2225
- Add events for packagers, repackagers, stock tickers, frogposts, postboxes, stock links (which aren't a peripheral yet) etc. The events would trigger when:
  - orders are made 
  - packages are made
  - packages are received

   Currently waiting for https://github.com/Creators-of-Create/Create/pull/7453 to get pulled in first.
- Cleaning up redstone requester's `.configureCraft()` to not get reset when the player interacts with it. Context: https://github.com/Creators-of-Create/Create/pull/7883#issuecomment-2848115787
- Adding requesting crafts with the stock ticker. Held back by the fact we're already testing our luck with how complex stock ticker's `requestFiltered` is so throwing more complexity at it sounds like a bad idea. I'd say after the redstone requester becomes 100% jank free, we can go about upping the stock ticker.
- Currently the tablecloth (very rarely) refuses to update the model on the client after we change it's wares via code (probs a block update thing). We've tried basically everything, got it working, now one of the updates in the last 3 months makes it break randomly again lol. If you guys can find a way to fix it that'd be lovely, if you can even manage to reproduce it.


As for the serial tests i'm working on, i'm just making a super minimal world that can hopefully be crunched to a megabyte or two that works off a lua library i made to take snapshots of peripherals, then runs some operations one after another.
It also clones the mutable block entities above down into itself to make re-running it super quick for debugging

I know there's more correct ways to do tests, esp since this is like insanely hard to automate, but ideally when it's done you guys can have the script run on startup and if it succeeds place a dummy block that a proper serial test checks for - that's out of my scope though. The code for the tests should work for whatever approach it's run through
I'll have the world dir on my github when it's done done
![tests](https://github.com/user-attachments/assets/c1183812-3390-493c-83b1-3eaf00a85762)
![image](https://github.com/user-attachments/assets/050d2de7-513e-442f-89fd-6beb1ca42d19)

So, i'm gonna open the pr, as the code's not gonna be changing much from now on.
Back to whacking at the wiki and tests at my snail's pace now :>

BirbIrl committed:
fixed linting for FrogPortBlockEntity

BirbIrl committed:
fixed linting on PackagerBlockEntity

BirbIrl committed:
added .getConfiguration and .setConfiguration to frogport and postbox

BirbIrl commented:
> The PR is now officially feature complete. Any new commit should be a bugfix/cleanup

I lied, added a new feature, now you can config the frogports to either `send and recieve`/`only send` with `.getConfiguration` and `.setConfiguration`. Reason being - i needed this for my serial tests :>

BirbIrl committed:
De-sillified the last commit. Reason: ate sandwich, randomly remembered how to code properly again.

BirbIrl committed:
small fix, jokes on me for not checking what filterChanged() does colon p

BirbIrl committed:
added .setConfiguration and .getConfiguration to RR
also had to refactor .configure and .configureCraft to .setRequest and
.setCraftingRequest for this

uthoffcyra commented:
Hey all! I've been watching this PR from the sidelines, I'm not familiar with Java but moreso with the Lua side of things. I know this is nearing completion, and I don't want to add any more burden to you all, but I wanted know if there was any integration formed with ComputerCraft's [event system](https://tweaked.cc/module/os.html#v:pullEvent)?

It could be extremely useful to have an event that fires whenever, say, a Frogport or Postbox recieves a new package. If a player is waiting for a package with specific materials, a computer connected to it could detect that, see that the contents within match, and perform some other function like sending a message to that player, or activating a redstone lamp signifying that, or whatever.

Another example could be sending an update when a stock link's inventory updates, so that a program doesn't repeatedly need to check for them— the code would only be run when a change is made.

Again, not familiar with the Java side of things, and I don't believe Create has any integration with this elsewhere, so I'm not sure how realistic something like this would be. Regardless, thank you for your time!

BirbIrl commented:
> I wanted know if there was any integration formed with ComputerCraft's [event system](https://tweaked.cc/module/os.html#v:pullEvent)?


oh yeah yup, just two messages ago, in the future plans section i've mentioned:
>  Add events for packagers, repackagers, stock tickers, frogposts, postboxes, stock links (which aren't a peripheral yet) etc. The events would trigger when:
> - orders are made
> - packages are made
> - packages are received
> - Currently waiting for https://github.com/Creators-of-Create/Create/pull/7453 to get pulled in first.

(i don't blame you for not seeing that either, pretty sure the discussion under this PR is longer the gentoo install guide)

> sending an update when a stock link's inventory updates

the network doesn't really know *when* inventories have items added to them. Every time you request a package, it checks if the corresponding items are available. (since y'know, you could have stock links attached to like 20 barrels, means you'd have to run 20 checks a tick for if any item was added)
You'd have to make designated import points with ledgers that log each item that goes in, and transmits it over the rednet

we will however *try* and make events for when any package is ordered in the future :>

so, this is planned, we're just waiting for elementW's event integration first, as to not double the implementation!

uthoffcyra commented:
> oh yeah yup, just two messages ago, in the future plans section i've mentioned:

Omg, I totally missed that I'm so sorry. Thanks for the clarification! The stock link limitation makes sense, and the ledger station doesnt sound like it would be too difficult of a work around anyways. Excited to see this PR eventually merge and see people make a lot of cool stuff with it. Great work, and good luck with the rest of the features!

BirbIrl committed:
requestFiltered() fix, now the filter mutates over a copy of itself.
previously the filter would break after comparing against the item,
defaulting to _mode = contains.

BirbIrl commented:
The test world is finished and available for anyone to try and contribute to when making changes to the create logistics compatiblity :D

right here: https://github.com/BirbIrl/Create-CC-Integration-Tests

For anyone that's part of the create team - do anything with it, it's licensed under MIT. It has served me well personally, letting me catch three bugs that previously went under the radar :>

I'm going to link the test world in the main post of this PR as well

BirbIrl committed:
made Packager and Repackager share their functionality because why not

BirbIrl committed:
`address` is no longer optional in `requestFiltered` for consistency

BirbIrl committed:
made redstone requester's `getConfiguration` return void because it
never returned false

BirbIrl committed:
reworked luaPackageObjects to work as snapshots

BirbIrl committed:
smol tweak with the new changes

BirbIrl committed:
fixed requestFiltered, pls don't look at this it's embarrassing

BirbIrl commented:
*after a good 3 months...*
## the PR is now super-duper done.


### We have added computercraft functionality for:
- [Re](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/repackager)/[Packager](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/packager)
- [Stock Ticker](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/stock-ticker)
  - [An abstract syntax tree system for filtered item requests](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/stock-ticker#requestFiltered) to go along with it. Thanks @MaxOrtGit for making my terrible algorithm actually good and like tripling its features.
- [Redstone Requester](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/redstone-requester)
- [Table Cloth](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/table-cloth)
  - Which can now dynamically change the items it sells and their price.
- [Frogports](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/package-frogport)/[Postboxes](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/postbox)

On top of that, [packages are now lua objects](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/package-object) available by `getItemDetail` or `getPackage` for packagers, which let you peek inside the package and the [encoded order](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/order-data-object), to make your own sorting system. if the package is held inside a packager it also lets you change the address it's shipped to.

There is a [serial test world with it's own custom tool suite](https://github.com/BirbIrl/Create-CC-Integration-Tests) which i made because i noticed some update on the dev branch subtly broke one of our functions - this makes spotting such cases really quick and simple :>
Additionally since it's under MIT, you guys can grab anything written/built there and add it to your test suite.

We've made a [wiki PR](https://github.com/Creators-of-Create/wiki/pull/4) as well, it's on my github pages available to look at [here](https://birbirl.github.io/wiki/users/cc-tweaked-integration/logistics/packager). It has full documentation of everything we've added

### Future plans we'd love to implement (well), but didn't get to:
- implement the generic inventory api handler for the stock ticker, so we can export items from it. Currently held back by jank on CC's end on adding `generic` capabilities to `IPeripheral`'s and nesting functions and general not-worth-the-effort-ness on my end, since it's possible, would just be a lot of rewriting. Related topic here https://github.com/cc-tweaked/CC-Tweaked/discussions/2225
- Add events for packagers, repackagers, stock tickers, frogposts, postboxes, stock links (which aren't a peripheral yet) etc. The events would trigger when:
  - orders are made 
  - packages are made
  - packages are received

   Currently waiting for https://github.com/Creators-of-Create/Create/pull/7453 to get pulled in first.
- Cleaning up redstone requester's `.configureCraft()` to not get reset when the player interacts with it. Context: https://github.com/Creators-of-Create/Create/pull/7883#issuecomment-2848115787
- Adding requesting crafts with the stock ticker. Held back by the fact we're already testing our luck with how complex stock ticker's `requestFiltered` is so throwing more complexity at it sounds like a bad idea. I'd say after the redstone requester becomes 100% jank free, we can go about upping the stock ticker.
- Currently the tablecloth (very rarely) refuses to update the model on the client after we change it's wares via code. Reproducing it is completely random and i have, after testing the pr for hours this week, only seen it happen **once**, and when i re-ran a test to see if it would happen again, it didn't. The tablecloth is it's own living entity with a soul and it's only purpose is to toy with us. We tried fixing it like 10 different times, thinking we got it working, only for it randomly break after a month. This thing is beyond us and we're leaving it how it is. Good enough.


My job here is done :D, I'd like to thank everyone that worked with me on this PR:
- @AsterAether, @Karotte128 and @McArctic for the tons of effort in setting the foundations of this PR over the first week of development, coding tons, discussing how we should do things and being part time janitors.
- @lucyamonster for chipping a couple good ideas and planting code that then grew into even more good stuff, as well as being another part time janitor
- @ShimmerGlass and @aearil for actually playing through a survival world with the PR, coding in a huuge chunk of functionality that we missed and fixing bugs
- @MaxOrtGit for implementing the most complex systems in this pr, like the package+orderdata data structures and re-implementing `requestFiltered` into something great, as well as always being there to help debug this PR in this last week's stretch

plus two smoller thank-you's to:
- @Electric131 for catching a missing functionality in `requestFiltered` and proposing a change we ended up re-implementing on top of MaxOrt's rework.
- @uthoffcyra for actually reading the wiki PR and adding a couple edits to it

\+ extra thanks for everyone being super chill and great to work with!  Please don't see this as ***my*** pr just because it says so on github, this was a pretty equal team effort!

BirbIrl commented:
> Happy to merge this as soon as you consider it ready.

oh yeah @simibubi this is the part where i @ you and mention that yes the pr is merge ready :D

BirbIrl committed:
Merge remote-tracking branch 'upstream/mc1.20.1/dev' into mc1.20.1/dev

BirbIrl commented:
Updated to match upstream, the test world passes all tests - apart from the bug i mentioned in the Hungry Frogs II commit comments https://github.com/Creators-of-Create/Create/commit/9d2127b8a8859110a9e499b40bfd4ccb61ff59bf#r161075530 where it yields negative, but after adding the missing `!` it passes :D

IThundxr commented:
Thanks to everyone who contributed to this PR, this will definitely be something exciting for the folks who enjoy using CC:T alongside Create

HKXingluo commented:
I can't use pushItems or pullItems, what should I do?

BirbIrl commented:
> I can't use pushItems or pullItems, what should I do?

what create version? what peripheral? does the command not do anything, or does it error?

HKXingluo commented:
> > 我不能用pushItems或pullItems，我该怎么办？
> 
> 哪个版本？什么外设？命令是没有反应，还是会出错？

https://github.com/Creators-of-Create/Create/issues/9699

 Create_Packager does not have these two APIs
